_posts/2024-08-06-how-to-talk-about-CHERI.markdown

```diff
@@ -0,0 +1,185 @@
+The software model defines what valid paths exist between the initial boot state and normal execution with you holding a pointer to a heap object, the hardware guarantees that some such path must have existed for you to hold that pointer.
+
+The two guarantees about immutability (code may not modify an object passed via immutable reference or any object any object reachable from an object that is passed as a deeply immutable reference) are simple properties once you have unforgeable CHERI capabilities.
+These are both enforced via *permissions*.
+Each capability carries a set of permissions that defines the set of things that it can be used for.
+Removing store permission means that it can't be used with store instructions (and so gives a read-only pointer).
+Morello and CHERIoT also defined a load-mutable permission.
+This permission allows you to transitively load capabilities that have store permission.
+When you load a capability, its store and load-mutable permissions are anded with load-mutable permission of the capability that authorised the load.
+This gives a very simple way of enforcing deep immutability.
+Removing a permission (and so constructing a deeply or shallowly immutable reference) is a simple register-register operation, less expensive in hardware than addition.
+
+The last property from the list, the ability to hand out tamper-proof opaque references, is implemented with CHERI's *sealing* mechanism.
+CHERI sealing associates an object type (a numerical value) with a pointer and, at the same time, makes the pointer unusable.
+Sealing requires a capability that authorises sealing with the specific type.
+Unsealing similarly requires a capability that authorises unsealing with a specific type.
+You can use this to enforce type safety for opaque types in the presence of unsafe code.
+
+I used this extensively in the [CHERI-JNI](https://www.cl.cam.ac.uk/research/security/ctsrd/pdfs/201704-asplos-cherijni.pdf) work back in 2017.
+When a Java program passed an object pointer to C, it was passed as a sealed capability.
+C code could do nothing with this other than pass it back to the Java VM via Java Native Interface (JNI) functions.
+The JNI exposes functions for getting or setting fields and calling methods.
+Each of these could unseal the object and know that it was a valid Java object.
+From there, there was no need for CHERI to be involved with type safety because every Java object carries a pointer to its class and so type checks were possible to implement purely in software.
+This highlights one of the core goals of CHERI: It *enables* languages to be safe, it doesn't *mandate* how they enforce that safety.
+The Java VM can efficiently enforce type safety (and therefore memory safety) within its own world without CHERI (though CHERI can improve performance in a few places), but CHERI enables it to retain these properties even when calling C.
+
+The remaining three properties are somewhat more specific to CHERIoT.
```

Comment:
I found it a bit hard to make the link at this point between "remaining three" and the list of things untrusted code can't do - maybe because we're covering them is a different order.   Maybe some sub headings based on the list items would help with the structure / flow ?